File: go/loader_test.go
package reflex

import (
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func fixtureDir() string {
	_, file, _, _ := runtime.Caller(0)
	return filepath.Join(filepath.Dir(file), "..", "docs", "fixtures")
}

func readFixture(t *testing.T, name string) []byte {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(fixtureDir(), name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	return data
}

func stubGuardFn(_ BlackboardReader) (bool, error) {
	return true, nil
}

// assertValidationError is defined in registry_test.go — reused here

// ---------------------------------------------------------------------------
// Valid inputs — fixtures
// ---------------------------------------------------------------------------

func TestLoadWorkflow_Greeting(t *testing.T) {
	wf, err := LoadWorkflow(readFixture(t, "greeting.json"), nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if wf.ID != "greeting" {
		t.Errorf("id = %q, want greeting", wf.ID)
	}
	if wf.Entry != "ASK_NAME" {
		t.Errorf("entry = %q, want ASK_NAME", wf.Entry)
	}
	if len(wf.Nodes) != 3 {
		t.Errorf("nodes = %d, want 3", len(wf.Nodes))
	}
	if len(wf.Edges) != 2 {
		t.Errorf("edges = %d, want 2", len(wf.Edges))
	}
	// No guards on any edge
	for _, e := range wf.Edges {
		if e.Guard != nil {
			t.Errorf("edge %s: expected no guard", e.ID)
		}
	}
}

func TestLoadWorkflow_DefinePhysicalObject(t *testing.T) {
	wf, err := LoadWorkflow(readFixture(t, "define-physical-object.json"), nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if wf.ID != "define-physical-object" {
		t.Errorf("id = %q", wf.ID)
	}
	if len(wf.Nodes) != 6 {
		t.Errorf("nodes = %d, want 6", len(wf.Nodes))
	}
	if len(wf.Edges) != 6 {
		t.Errorf("edges = %d, want 6", len(wf.Edges))
	}

	// Check exists guard on e-branch-to-part
	var branchToPart *Edge
	for i := range wf.Edges {
		if wf.Edges[i].ID == "e-branch-to-part" {
			branchToPart = &wf.Edges[i]
			break
		}
	}
	if branchToPart == nil {
		t.Fatal("edge e-branch-to-part not found")
	}
	bg, ok := branchToPart.Guard.(*BuiltinGuard)
	if !ok {
		t.Fatalf("guard is %T, want *BuiltinGuard", branchToPart.Guard)
	}
	if bg.Type != GuardExists || bg.Key != "needsPart" {
		t.Errorf("guard = {%s, %s}, want {exists, needsPart}", bg.Type, bg.Key)
	}

	// Check not-exists guard on e-branch-to-spec
	var branchToSpec *Edge
	for i := range wf.Edges {
		if wf.Edges[i].ID == "e-branch-to-spec" {
			branchToSpec = &wf.Edges[i]
			break
		}
	}
	if branchToSpec == nil {
		t.Fatal("edge e-branch-to-spec not found")
	}
	bg2, ok := branchToSpec.Guard.(*BuiltinGuard)
	if !ok {
		t.Fatalf("guard is %T, want *BuiltinGuard", branchToSpec.Guard)
	}
	if bg2.Type != GuardNotExists || bg2.Key != "needsPart" {
		t.Errorf("guard = {%s, %s}, want {not-exists, needsPart}", bg2.Type, bg2.Key)
	}
}

func TestLoadWorkflow_DefinePartObject(t *testing.T) {
	wf, err := LoadWorkflow(readFixture(t, "define-part-object.json"), nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if wf.ID != "define-part-object" {
		t.Errorf("id = %q", wf.ID)
	}
	if len(wf.Nodes) != 3 {
		t.Errorf("nodes = %d, want 3", len(wf.Nodes))
	}
	if len(wf.Edges) != 2 {
		t.Errorf("edges = %d, want 2", len(wf.Edges))
	}
}

func TestLoadWorkflow_InvocationSpec(t *testing.T) {
	wf, err := LoadWorkflow(readFixture(t, "define-physical-object.json"), nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	definePart := wf.Nodes["DEFINE_PART"]
	if definePart == nil {
		t.Fatal("node DEFINE_PART not found")
	}
	if definePart.Invokes == nil {
		t.Fatal("DEFINE_PART.Invokes is nil")
	}
	if definePart.Invokes.WorkflowID != "define-part-object" {
		t.Errorf("workflowId = %q", definePart.Invokes.WorkflowID)
	}
	if len(definePart.Invokes.ReturnMap) != 1 {
		t.Fatalf("returnMap len = %d, want 1", len(definePart.Invokes.ReturnMap))
	}
	rm := definePart.Invokes.ReturnMap[0]
	if rm.ParentKey != "Part Concept" || rm.ChildKey != "partConcept" {
		t.Errorf("returnMap[0] = {%q, %q}", rm.ParentKey, rm.ChildKey)
	}
}

func TestLoadWorkflow_NodeSpec(t *testing.T) {
	wf, err := LoadWorkflow(readFixture(t, "greeting.json"), nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	askName := wf.Nodes["ASK_NAME"]
	if askName == nil {
		t.Fatal("node ASK_NAME not found")
	}
	if askName.Spec["prompt"] != "Ask the user for their name" {
		t.Errorf("spec.prompt = %v", askName.Spec["prompt"])
	}
	if askName.Spec["outputKey"] != "userName" {
		t.Errorf("spec.outputKey = %v", askName.Spec["outputKey"])
	}
}

// ---------------------------------------------------------------------------
// Custom guard resolution
// ---------------------------------------------------------------------------

func TestLoadWorkflow_CustomGuardResolved(t *testing.T) {
	data := []byte(`{
		"id": "test",
		"entry": "A",
		"nodes": {
			"A": {"id": "A", "spec": {}},
			"B": {"id": "B", "spec": {}}
		},
		"edges": [{
			"id": "e1", "from": "A", "to": "B", "event": "NEXT",
			"guard": {"type": "custom", "name": "myGuard"}
		}]
	}`)

	guard := &CustomGuardFunc{Fn: stubGuardFn}
	wf, err := LoadWorkflow(data, &LoadWorkflowOptions{
		Guards: GuardRegistry{"myGuard": guard},
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if wf.Edges[0].Guard != guard {
		t.Error("guard not resolved from registry")
	}
}

func TestLoadWorkflow_CustomGuardMissing(t *testing.T) {
	data := []byte(`{
		"id": "test",
		"entry": "A",
		"nodes": {
			"A": {"id": "A", "spec": {}},
			"B": {"id": "B", "spec": {}}
		},
		"edges": [{
			"id": "e1", "from": "A", "to": "B", "event": "NEXT",
			"guard": {"type": "custom", "name": "nonExistent"}
		}]
	}`)
	_, err := LoadWorkflow(data, nil)
	assertValidationError(t, err, ErrUnknownGuardReference)
}

func TestLoadWorkflow_CustomGuardMissingFromRegistry(t *testing.T) {
	data := []byte(`{
		"id": "test",
		"entry": "A",
		"nodes": {
			"A": {"id": "A", "spec": {}},
			"B": {"id": "B", "spec": {}}
		},
		"edges": [{
			"id": "e1", "from": "A", "to": "B", "event": "NEXT",
			"guard": {"type": "custom", "name": "missing"}
		}]
	}`)
	guard := &CustomGuardFunc{Fn: stubGuardFn}
	_, err := LoadWorkflow(data, &LoadWorkflowOptions{
		Guards: GuardRegistry{"otherGuard": guard},
	})
	assertValidationError(t, err, ErrUnknownGuardReference)
}

// ---------------------------------------------------------------------------
// Schema violations
// ---------------------------------------------------------------------------

func TestLoadWorkflow_InvalidJSON(t *testing.T) {
	_, err := LoadWorkflow([]byte("not valid json{"), nil)
	assertValidationError(t, err, ErrSchemaViolation)
}

func TestLoadWorkflow_MissingID(t *testing.T) {
	data := []byte(`{"entry":"A","nodes":{"A":{"id":"A","spec":{}}},"edges":[]}`)
	_, err := LoadWorkflow(data, nil)
	assertValidationError(t, err, ErrSchemaViolation)
}

func TestLoadWorkflow_MissingEntry(t *testing.T) {
	data := []byte(`{"id":"x","nodes":{"A":{"id":"A","spec":{}}},"edges":[]}`)
	_, err := LoadWorkflow(data, nil)
	assertValidationError(t, err, ErrSchemaViolation)
}

func TestLoadWorkflow_MissingNodes(t *testing.T) {
	data := []byte(`{"id":"x","entry":"A","edges":[]}`)
	_, err := LoadWorkflow(data, nil)
	assertValidationError(t, err, ErrSchemaViolation)
}

func TestLoadWorkflow_MissingEdges(t *testing.T) {
	data := []byte(`{"id":"x","entry":"A","nodes":{"A":{"id":"A","spec":{}}}}`)
	_, err := LoadWorkflow(data, nil)
	assertValidationError(t, err, ErrSchemaViolation)
}

func TestLoadWorkflow_UnknownGuardType(t *testing.T) {
	data := []byte(`{
		"id": "x",
		"entry": "A",
		"nodes": {
			"A": {"id": "A", "spec": {}},
			"B": {"id": "B", "spec": {}}
		},
		"edges": [{
			"id": "e1", "from": "A", "to": "B", "event": "NEXT",
			"guard": {"type": "bogus", "key": "x"}
		}]
	}`)
	_, err := LoadWorkflow(data, nil)
	assertValidationError(t, err, ErrSchemaViolation)
}

func TestLoadWorkflow_SchemaFieldAllowed(t *testing.T) {
	// $schema field should be silently ignored
	wf, err := LoadWorkflow(readFixture(t, "greeting.json"), nil)
	if err != nil {
		t.Fatalf("$schema field should be allowed: %v", err)
	}
	if wf.ID != "greeting" {
		t.Errorf("id = %q", wf.ID)
	}
}

File: typescript/src/loader.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { loadWorkflow } from './loader';
import { WorkflowValidationError } from './registry';
import type { BuiltinGuard, CustomGuard, BlackboardReader } from './types';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const __dirname = dirname(fileURLToPath(import.meta.url));
const fixtureDir = resolve(__dirname, '../../docs/fixtures');

function readFixture(name: string): string {
  return readFileSync(resolve(fixtureDir, name), 'utf-8');
}

function parseFixture(name: string): unknown {
  return JSON.parse(readFixture(name));
}

const stubGuard = (_bb: BlackboardReader) => true;

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('loadWorkflow', () => {
  // -----------------------------------------------------------------------
  // Valid inputs
  // -----------------------------------------------------------------------

  describe('valid inputs', () => {
    it('loads greeting.json from parsed object', () => {
      const wf = loadWorkflow(parseFixture('greeting.json'));
      expect(wf.id).toBe('greeting');
      expect(wf.entry).toBe('ASK_NAME');
      expect(Object.keys(wf.nodes)).toHaveLength(3);
      expect(wf.edges).toHaveLength(2);
    });

    it('loads greeting.json from raw JSON string', () => {
      const wf = loadWorkflow(readFixture('greeting.json'));
      expect(wf.id).toBe('greeting');
      expect(Object.keys(wf.nodes)).toHaveLength(3);
    });

    it('loads define-part-object.json with invocation spec', () => {
      const wf = loadWorkflow(parseFixture('define-part-object.json'));
      expect(wf.id).toBe('define-part-object');
      expect(wf.entry).toBe('PART_CLASSIFY');
      expect(Object.keys(wf.nodes)).toHaveLength(3);
      expect(wf.edges).toHaveLength(2);
    });

    it('loads define-physical-object.json with builtin guards', () => {
      const wf = loadWorkflow(
        parseFixture('define-physical-object.json'),
      );
      expect(wf.id).toBe('define-physical-object');
      expect(Object.keys(wf.nodes)).toHaveLength(6);
      expect(wf.edges).toHaveLength(6);

      // Check exists guard
      const branchToPart = wf.edges.find(
        (e) => e.id === 'e-branch-to-part',
      )!;
      expect(branchToPart.guard).toBeDefined();
      const existsGuard = branchToPart.guard as BuiltinGuard;
      expect(existsGuard.type).toBe('exists');
      expect(existsGuard.key).toBe('needsPart');

      // Check not-exists guard
      const branchToSpec = wf.edges.find(
        (e) => e.id === 'e-branch-to-spec',
      )!;
      expect(branchToSpec.guard).toBeDefined();
      const notExistsGuard = branchToSpec.guard as BuiltinGuard;
      expect(notExistsGuard.type).toBe('not-exists');
      expect(notExistsGuard.key).toBe('needsPart');
    });

    it('preserves invocation spec on nodes', () => {
      const wf = loadWorkflow(
        parseFixture('define-physical-object.json'),
      );
      const definePart = wf.nodes['DEFINE_PART'];
      expect(definePart.invokes).toBeDefined();
      expect(definePart.invokes!.workflowId).toBe('define-part-object');
      expect(definePart.invokes!.returnMap).toEqual([
        { parentKey: 'Part Concept', childKey: 'partConcept' },
      ]);
    });

    it('preserves node spec as freeform object', () => {
      const wf = loadWorkflow(parseFixture('greeting.json'));
      const askName = wf.nodes['ASK_NAME'];
      expect(askName.spec).toEqual({
        prompt: 'Ask the user for their name',
        outputKey: 'userName',
      });
    });

    it('preserves node description when present', () => {
      const wf = loadWorkflow(parseFixture('greeting.json'));
      expect(wf.nodes['GREET'].description).toBe(
        'Generate a personalized greeting',
      );
    });

    it('allows $schema field in input', () => {
      // Fixtures include $schema — should not throw
      const wf = loadWorkflow(parseFixture('greeting.json'));
      expect(wf.id).toBe('greeting');
    });

    it('edges without guards have guard undefined', () => {
      const wf = loadWorkflow(parseFixture('greeting.json'));
      for (const edge of wf.edges) {
        expect(edge.guard).toBeUndefined();
      }
    });
  });

  // -----------------------------------------------------------------------
  // Custom guard resolution
  // -----------------------------------------------------------------------

  describe('custom guard resolution', () => {
    it('resolves custom guard from registry', () => {
      const input = {
        id: 'test',
        entry: 'A',
        nodes: {
          A: { id: 'A', spec: {} },
          B: { id: 'B', spec: {} },
        },
        edges: [
          {
            id: 'e1',
            from: 'A',
            to: 'B',
            event: 'NEXT',
            guard: { type: 'custom', name: 'myGuard' },
          },
        ],
      };

      const wf = loadWorkflow(input, { guards: { myGuard: stubGuard } });
      const guard = wf.edges[0].guard as CustomGuard;
      expect(guard.type).toBe('custom');
      expect(guard.evaluate).toBe(stubGuard);
    });

    it('throws UNKNOWN_GUARD_REFERENCE for missing guard', () => {
      const input = {
        id: 'test',
        entry: 'A',
        nodes: {
          A: { id: 'A', spec: {} },
          B: { id: 'B', spec: {} },
        },
        edges: [
          {
            id: 'e1',
            from: 'A',
            to: 'B',
            event: 'NEXT',
            guard: { type: 'custom', name: 'nonExistent' },
          },
        ],
      };

      expect(() => loadWorkflow(input)).toThrow(WorkflowValidationError);
      try {
        loadWorkflow(input);
      } catch (e) {
        const err = e as WorkflowValidationError;
        expect(err.code).toBe('UNKNOWN_GUARD_REFERENCE');
        expect(err.workflowId).toBe('test');
      }
    });

    it('throws UNKNOWN_GUARD_REFERENCE when registry provided but name missing', () => {
      const input = {
        id: 'test',
        entry: 'A',
        nodes: {
          A: { id: 'A', spec: {} },
          B: { id: 'B', spec: {} },
        },
        edges: [
          {
            id: 'e1',
            from: 'A',
            to: 'B',
            event: 'NEXT',
            guard: { type: 'custom', name: 'missing' },
          },
        ],
      };

      expect(() =>
        loadWorkflow(input, { guards: { otherGuard: stubGuard } }),
      ).toThrow(WorkflowValidationError);
    });
  });

  // -----------------------------------------------------------------------
  // Schema violations
  // -----------------------------------------------------------------------

  describe('schema violations', () => {
    it('throws on invalid JSON string', () => {
      expect(() => loadWorkflow('not valid json{')).toThrow(
        WorkflowValidationError,
      );
      try {
        loadWorkflow('not valid json{');
      } catch (e) {
        expect((e as WorkflowValidationError).code).toBe('SCHEMA_VIOLATION');
      }
    });

    it('throws on missing id', () => {
      const input = { entry: 'A', nodes: { A: { id: 'A', spec: {} } }, edges: [] };
      expect(() => loadWorkflow(input)).toThrow(WorkflowValidationError);
    });

    it('throws on missing entry', () => {
      const input = { id: 'x', nodes: { A: { id: 'A', spec: {} } }, edges: [] };
      expect(() => loadWorkflow(input)).toThrow(WorkflowValidationError);
    });

    it('throws on missing nodes', () => {
      const input = { id: 'x', entry: 'A', edges: [] };
      expect(() => loadWorkflow(input)).toThrow(WorkflowValidationError);
    });

    it('throws on missing edges', () => {
      const input = { id: 'x', entry: 'A', nodes: { A: { id: 'A', spec: {} } } };
      expect(() => loadWorkflow(input)).toThrow(WorkflowValidationError);
    });

    it('throws when nodes is an array instead of object', () => {
      const input = {
        id: 'x',
        entry: 'A',
        nodes: [{ id: 'A', spec: {} }],
        edges: [],
      };
      expect(() => loadWorkflow(input)).toThrow(WorkflowValidationError);
    });

    it('throws on unknown guard type', () => {
      const input = {
        id: 'x',
        entry: 'A',
        nodes: {
          A: { id: 'A', spec: {} },
          B: { id: 'B', spec: {} },
        },
        edges: [
          {
            id: 'e1',
            from: 'A',
            to: 'B',
            event: 'NEXT',
            guard: { type: 'bogus', key: 'x' },
          },
        ],
      };
      expect(() => loadWorkflow(input)).toThrow(WorkflowValidationError);
    });

    it('throws on non-object input', () => {
      expect(() => loadWorkflow(42)).toThrow(WorkflowValidationError);
    });

    it('loads node with inputs and outputs contracts', () => {
      const wf = loadWorkflow({
        id: 'contracts',
        entry: 'A',
        nodes: {
          A: {
            id: 'A',
            spec: {},
            inputs: [
              { key: 'userName', required: true, description: 'The user name' },
              { key: 'optional', required: false },
            ],
            outputs: [
              { key: 'greeting', guaranteed: true, description: 'The greeting message' },
            ],
          },
        },
        edges: [],
      });
      expect(wf.nodes['A'].inputs).toHaveLength(2);
      expect(wf.nodes['A'].inputs![0]).toEqual({
        key: 'userName',
        required: true,
        description: 'The user name',
      });
      expect(wf.nodes['A'].inputs![1]).toEqual({
        key: 'optional',
        required: false,
      });
      expect(wf.nodes['A'].outputs).toHaveLength(1);
      expect(wf.nodes['A'].outputs![0]).toEqual({
        key: 'greeting',
        guaranteed: true,
        description: 'The greeting message',
      });
    });

    it('nodes without contracts load with undefined inputs/outputs', () => {
      const wf = loadWorkflow({
        id: 'no-contracts',
        entry: 'A',
        nodes: { A: { id: 'A', spec: {} } },
        edges: [],
      });
      expect(wf.nodes['A'].inputs).toBeUndefined();
      expect(wf.nodes['A'].outputs).toBeUndefined();
    });

    it('all schema violations use SCHEMA_VIOLATION code', () => {
      const cases = [
        { entry: 'A', nodes: {}, edges: [] }, // missing id
        'not json{', // invalid JSON string
        42, // non-object
      ];

      for (const input of cases) {
        try {
          loadWorkflow(input);
        } catch (e) {
          expect((e as WorkflowValidationError).code).toBe('SCHEMA_VIOLATION');
        }
      }
    });
  });
});
